src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs

```diff
@@ -0,0 +1,701 @@
+        }
+
+        public int Compare(int x, int y)
+        {
+            throw new NotSupportedException(NullComparerExceptionMessage);
+        }
+    }
+
+    private struct EntryComparerByDouble : IEntryComparer, IComparer<UnmanagedSpan>
+    {
+        private int _comparerId;
+        private Lookup<Int64LookupKey> _lookup;
+        private UnmanagedSpan<long> _batchResults;
+
+        public void SortBatch<TComparer2, TComparer3>(ref SortingMultiMatch<TInner> match, LowLevelTransaction llt, PageLocator pageLocator,
+            UnmanagedSpan<long> batchResults, Span<long> batchTermIds,
+            UnmanagedSpan* batchTerms, OrderMetadata[] orderMetadata, TComparer2 comparer2, TComparer3 comparer3)
+            where TComparer2 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+            where TComparer3 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+        {
+            if (_lookup == null) // field does not exist, so arbitrary sort order, whatever query said goes
+            {
+                match._results.Add(batchResults);
+                return;
+            }
+
+            _lookup.GetFor(batchResults, batchTermIds, BitConverter.DoubleToInt64Bits(double.MinValue));
+            var indexes = EntryComparerHelper.NumericSortBatch(ref match, batchTermIds, batchTerms, new EntryComparerByDouble(), comparer2, comparer3);
+            for (int i = 0; i < indexes.Length; i++)
+            {
+                match._results.Add(batchResults[indexes[i]]);
+            }
+        }
+
+        public Slice GetSortFieldName(ref SortingMultiMatch<TInner> match)
+        {
+            IndexFieldsMappingBuilder.GetFieldNameForDoubles(match._searcher.Allocator, match._orderMetadata[_comparerId].Field.FieldName, out var dblName);
+            return dblName;
+        }
+
+        public void Init(ref SortingMultiMatch<TInner> match, UnmanagedSpan<long> batchResults, int comparerId)
+        {
+            _comparerId = comparerId;
+            _lookup = match._searcher.EntriesToTermsReader(GetSortFieldName(ref match));
+            _batchResults = batchResults;
+        }
+
+        public int Compare(UnmanagedSpan x, UnmanagedSpan y)
+        {
+            return x.Double.CompareTo(y.Double);
+        }
+
+        public int Compare(int x, int y)
+        {
+            if (_lookup == null) // field does not exist, so arbitrary sort order, whatever query said goes
+                return 0;
+
+            Span<long> buffer = stackalloc long[4] {_batchResults[x], _batchResults[y], -1, -1};
+            var swap = buffer[0] > buffer[1];
+            if (swap)
+                (buffer[0], buffer[1]) = (buffer[1], buffer[0]);
+            
+            _lookup.GetFor(buffer[..2], buffer[2..], BitConverter.DoubleToInt64Bits(double.MinValue));
+            return buffer[swap ? 3 : 2].CompareTo(buffer[swap ? 2 : 3]);
+        }
+    }
```

Comment:
@ayende do that long keys persist in order in case of double or do I've to convert them into double?

Reply:
You have to compare them using doubles, not long
The bit patterns are the same, the meaning is not